src/v/cluster/tm_tx_hash_ranges.h

```diff
@@ -0,0 +1,301 @@
+/*
+ * Copyright 2023 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+
+#include "hashing/murmur.h"
+#include "kafka/protocol/types.h"
+#include "model/fundamental.h"
+#include "reflection/adl.h"
+
+#include <cstdint>
+#include <vector>
+
+namespace cluster {
+
+static uint32_t get_default_range_size(int32_t tm_partitions_amount) {
+    // integer division acts as div
+    // size of last range can be larger than default range size
+    // last range covers hashes up to UINT32_MAX
+    return UINT32_MAX / tm_partitions_amount;
+}
+
+// Return hash of tx_id in range [0, UINT32_MAX]
+inline uint32_t get_tx_id_hash(const kafka::transactional_id& tx_id) {
+    return murmur2(ss::sstring(tx_id).data(), ss::sstring(tx_id).size());
+}
+
+enum class tm_tx_hash_ranges_errc {
+    success = 0,
+    not_hosted = 1,
+    intersection = 2
+};
+
+// Defines hash range as section where both ends are included
+struct tm_hash_range
+  : serde::
+      envelope<tm_hash_range, serde::version<0>, serde::compat_version<0>> {
+    uint32_t first;
+    uint32_t last;
+
+    tm_hash_range() = default;
+
+    tm_hash_range(uint32_t f, uint32_t l)
+      : first(f)
+      , last(l) {}
+
+    auto serde_fields() { return std::tie(first, last); }
+
+    bool contains(uint32_t v) const { return v >= first && v <= last; }
+
+    bool contains(tm_hash_range r) const {
+        return r.first >= first && r.last <= last;
+    }
+
+    bool intersects(tm_hash_range r) const {
+        return (r.first >= first && r.first <= last)
+               || (r.last >= first && r.last <= last) || r.contains(*this);
+    }
+};
+
+inline tm_hash_range default_tm_hash_range(
+  model::partition_id tm_partition, int32_t partitions_amount) {
+    uint32_t range_size = get_default_range_size(partitions_amount);
+    uint32_t hash_range_begin = range_size * tm_partition;
+    uint32_t hash_range_end = (tm_partition + 1) == partitions_amount
+                                ? UINT32_MAX
```

Comment:
can we use a constexpr variable that we control rather than uint32_max ?

Reply:
with uint32_max it's pretty obvious what it means. what are you suggesting to use instead of it? `numeric_limits<uint32_t>::max()`

Reply:
yeh just like `constexpr auto hash_max = numeric_limits<uint32_t>::max()` or something. it's not a big deal either way, it's just that `UINT32_MAX` is a very C-ish thing